### ISSUE TITLE
to_s method called for skipped resources

### ISSUE THREAD

**Issue by alexpop** [inspec/inspec#4048]
Given a profile(with no `supports` defined in inspec.yml) with the following control:

```
control "registry-test" do
  describe registry_key("HKEY_LOCAL_MACHINE\\Software\\Microsoft\\Windows\\CurrentVersion\\Policies\\System") do
    it { should have_property "NoConnectedUser" }
  end
end
```

When running against a non-windows node it will raise the following exception:

```
NoMethodError: undefined method `[]' for nil:NilClass
  /Users/apop/git/inspec/lib/resources/registry_key.rb:125:in `to_s'
  /opt/chefdk/embedded/lib/ruby/gems/2.5.0/gems/rspec-core-3.8.0/lib/rspec/core/metadata.rb:180:in `build_description_from'
  /opt/chefdk/embedded/lib/ruby/gems/2.5.0/gems/rspec-core-3.8.0/lib/rspec/core/metadata.rb:133:in `populate'
  /opt/chefdk/embedded/lib/ruby/gems/2.5.0/gems/rspec-core-3.8.0/lib/rspec/core/metadata.rb:258:in `create'
  /opt/chefdk/embedded/lib/ruby/gems/2.5.0/gems/rspec-core-3.8.0/lib/rspec/core/example_group.rb:422:in `set_it_up'
  /opt/chefdk/embedded/lib/ruby/gems/2.5.0/gems/rspec-core-3.8.0/lib/rspec/core/example_group.rb:386:in `subclass'
  /opt/chefdk/embedded/lib/ruby/gems/2.5.0/gems/rspec-core-3.8.0/lib/rspec/core/example_group.rb:260:in `block in define_example_group_method'
  /Users/apop/git/inspec/lib/inspec/runner_rspec.rb:27:in `example_group'
  /Users/apop/git/inspec/lib/inspec/runner.rb:287:in `rspec_failed_block'
  /Users/apop/git/inspec/lib/inspec/runner.rb:261:in `get_check_example'
  /Users/apop/git/inspec/lib/inspec/runner.rb:273:in `block in register_rule'
  /Users/apop/git/inspec/lib/inspec/runner.rb:272:in `each'
  /Users/apop/git/inspec/lib/inspec/runner.rb:272:in `flat_map'
  /Users/apop/git/inspec/lib/inspec/runner.rb:272:in `register_rule'
  /Users/apop/git/inspec/lib/inspec/runner.rb:115:in `block in load'
  /Users/apop/git/inspec/lib/inspec/runner.rb:114:in `each'
  /Users/apop/git/inspec/lib/inspec/runner.rb:114:in `load'
  /Users/apop/git/inspec/lib/inspec/runner.rb:121:in `run'
  /Users/apop/git/inspec/lib/inspec/cli.rb:269:in `exec'
  /opt/chefdk/embedded/lib/ruby/gems/2.5.0/gems/thor-0.20.3/lib/thor/command.rb:27:in `run'
  /opt/chefdk/embedded/lib/ruby/gems/2.5.0/gems/thor-0.20.3/lib/thor/invocation.rb:126:in `invoke_command'
  /opt/chefdk/embedded/lib/ruby/gems/2.5.0/gems/thor-0.20.3/lib/thor.rb:387:in `dispatch'
  /opt/chefdk/embedded/lib/ruby/gems/2.5.0/gems/thor-0.20.3/lib/thor/base.rb:466:in `start'
  /Users/apop/git/inspec/inspec-bin/bin/inspec:15:in `<top (required)>'
  /opt/chefdk/embedded/lib/ruby/gems/2.5.0/bin/inspec:23:in `load'
  /opt/chefdk/embedded/lib/ruby/gems/2.5.0/bin/inspec:23:in `<top (required)>'
```

This is because the `registry_key` resource is skipped due to an unsupported platform.
The [to_s method](https://github.com/inspec/inspec/blob/v4.4.3/lib/resources/registry_key.rb#L124) of `registry_key` is being called without the resource being initialized. Because of this, the method runs with the `@options` hash being nil.

If I change the `to_s` method of `registry_key` like this:
```ruby
def to_s
  @options ||= {}
  "Registry Key #{@options[:name]}"
end
```

The exception goes away and the following control failure is displayed in the CLI:
```
  ×  registry-test: Registry Key
     ×  Registry Key
     Resource `registry_key` is not supported on platform mac_os_x/17.7.0.
```

I checked and at least 10 more core inspec resources rely on initialized variables in their`to_s` methods.

We can update those resources like I did with the `registry_key` one or not call `to_s` when a resource is skipped. And have an output like this:
```
  ×  registry-test:
     ×  Resource `registry_key` is not supported on platform mac_os_x/17.7.0.
```

**Comment by aaronlippold** [comment]
I agree it's much cleaner to the end user.

**Comment by zenspider** [comment]
I'm working on #4214 right now and it changes the above repro to:

```
8462 % ./run.sh exec bug4048.rb

Profile: tests from bug4048.rb (tests from bug4048.rb)
Version: (not specified)
Target:  local://

  ↺  registry-test: Registry Key HKEY_LOCAL_MACHINE\Software\Microsoft\Windows\CurrentVersion\Policies\System
     ↺  The `registry_key` resource is not supported on your OS yet.


Profile Summary: 0 successful controls, 0 control failures, 1 control skipped
Test Summary: 0 successful, 0 failures, 1 skipped
```

**Comment by zenspider** [comment]
Or:

```
8473 % ./run.sh exec bug4048.rb

Profile: tests from bug4048.rb (tests from bug4048.rb)
Version: (not specified)
Target:  local://

  ×  registry-test: bug4048.rb:1
     ×  Control Source Code Error bug4048.rb:1 
     Unsupported resource/backend combination: registry_key / mac_os_x. Exiting.


Profile Summary: 0 successful controls, 1 control failure, 0 controls skipped
Test Summary: 0 successful, 1 failure, 0 skipped
```

**Comment by zenspider** [comment]
I'm trying to figure out the correct trade-off right now. Both forms break a test we currently have, but might be worth it... or there might be a more correct solution somewhere above this.

**Comment by zenspider** [comment]
Fixed by #4366
